src/DataStructures/ModalVector.hpp

```diff
@@ -0,0 +1,109 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "DataStructures/VectorImpl.hpp"
+#include "Utilities/PointerVector.hpp"  // IWYU pragma: keep
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_include <blaze/math/expressions/DVecMapExpr.h>
+// IWYU pragma: no_include <blaze/math/typetraits/IsVector.h>
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A class for storing spectral coefficients on a spectral grid.
+ *
+ * \details A ModalVector holds an array of spectral coefficients, and can be
+ * either owning (the array is deleted when the ModalVector goes out of scope)
+ * or non-owning, meaning it just has a pointer to an array.
+ *
+ * Only basic mathematical operations are supported with
+ * ModalVectors. ModalVectors may be added or subtracted, and the following
+ * unary operations are supported:
+ * - abs
+ * - max
+ * - min
+ *
+ * Also, addition, subtraction, multiplication, and division are supported with
+ * doubles and `ModalVector`s and multiplication is supported with
+ * `DiagonalModalOperator`s and `ModalVector`s.
+ */
+class ModalVector : public VectorImpl<double, ModalVector> {
+ public:
+  using VectorImpl<double, ModalVector>::operator=;
+  using VectorImpl<double, ModalVector>::VectorImpl;
+
+  MAKE_MATH_ASSIGN_EXPRESSION_POINTERVECTOR(+=, ModalVector)
+  MAKE_MATH_ASSIGN_EXPRESSION_POINTERVECTOR(-=, ModalVector)
+};
+
+namespace blaze {
+template <>
+struct IsVector<ModalVector> : std::true_type {};
+template <>
+struct TransposeFlag<ModalVector> : BoolConstant<ModalVector::transpose_flag> {
+};
+BLAZE_TRAIT_SPECIALIZE_BINARY_TRAIT(ModalVector, AddTrait);
+BLAZE_TRAIT_SPECIALIZE_BINARY_TRAIT(ModalVector, SubTrait);
+BLAZE_TRAIT_SPECIALIZE_COMPATIBLE_BINARY_TRAIT(ModalVector, double, MultTrait);
+BLAZE_TRAIT_SPECIALIZE_COMPATIBLE_BINARY_TRAIT(ModalVector, double, DivTrait);
```

Comment:
does this only allow division by double?

Reply:
The macro permits either ordering, so I'll revise it to specialize only the appropriate operation (`ModalVector` divided by `double`s).